cloudferrylib/os/identity/keystone.py

```diff
@@ -474,7 +477,57 @@ def _get_user_tenants_roles(self, tenant_list=None, user_list=None):
             tenant_list = []
         if user_list is None:
             user_list = []
-        user_tenants_roles = {}
+        if not self.config.identity.optimize_user_role_fetch:
+            user_tenants_roles = self._get_user_tenants_roles_by_api(tenant_list,
+                                                                     user_list)
+        else:
+            user_tenants_roles = self._get_user_tenants_roles_by_db(tenant_list,
+                                                                    user_list)
+        return user_tenants_roles
+
+    def _get_user_roles_cached(self):
+        all_roles = {}
+        if self.config.identity.optimize_user_role_fetch:
+            res = self.mysql_connector.execute(
+                "SELECT * FROM user_project_metadata")
+            for row in res:
+                user_id = row[0]
+                tenant_id = row[1]
+                roles_ids = ast.literal_eval(row[2])['roles']
+                all_roles[user_id] = {} if not user_id in all_roles else all_roles[user_id]
+                all_roles[user_id][tenant_id] = [] \
+                    if not tenant_id in all_roles[user_id] else all_roles[user_id][tenant_id]
+                all_roles[user_id][tenant_id].append(roles_ids)
+
+        def _get_user_roles(user_id, tenant_id):
+            if not self.config.identity.optimize_user_role_fetch:
+                roles = self.roles_for_user(user_id, tenant_id)
+            else:
+                roles = getattr(getattr(all_roles, user_id, {}), tenant_id, {})
+            return roles
+        return _get_user_roles
+
+    def _get_user_tenants_roles_by_db(self, tenant_list, user_list):
+        user_tenants_roles = {u.name: {t.name: [] for t in tenant_list}
+                              for u in user_list}
+        tenant_ids = {tenant.id: tenant.name for tenant in tenant_list}
+        user_ids = {user.id: user.name for user in user_list}
+        roles = {r.id: r for r in self.get_roles_list()}
+        res = self.mysql_connector.execute(
+            "SELECT * FROM user_project_metadata")
+        for row in res:
+            user_id = row[0]
+            tenant_id = row[1]
+            roles_ids = ast.literal_eval(row[2])['roles']
+            user_tenants_roles[user_ids[user_id]][tenant_ids[tenant_id]] = [{'role':
```

Comment:
Few other things to fix:
 - This line does not conform to pep8;
 - The loop could be rewritten to be much more readable (`connector.execute` returns tuple), also `row` is not really row, but column ;)
```
for user_id, tenant_id, roles in res:
   ...
```